network/Message.cpp

```diff
@@ -85,15 +94,74 @@ char* Message::Data() noexcept
 void Message::Swap(Message& rhs) noexcept {
     std::swap(m_type, rhs.m_type);
     std::swap(m_message_size, rhs.m_message_size);
+    std::swap(m_is_compressed, rhs.m_is_compressed);
     std::swap(m_message_text, rhs.m_message_text);
 }
 
 void Message::Reset() noexcept {
     m_type = MessageType::UNDEFINED;
     m_message_size = 0;
+    m_is_compressed = false;
     m_message_text.clear();
 }
 
+void Message::Compress() noexcept {
+    if (!m_is_compressed) {
+        std::string_view::size_type temp = m_message_size;
+        m_message_text = CompressImpl(m_message_text);
+        m_message_size = m_message_text.size();
+        m_is_compressed = true;
+    }
+}
+
+std::string Message::CompressImpl(const std::string& data) {
+    boost::iostreams::basic_array_source<char> src(data.c_str(), data.size());
+    boost::iostreams::stream<boost::iostreams::basic_array_source<char>> s_src(src);
+
+    std::string compressed;
+    compressed.reserve(std::pow(2.0, 26.0));
+
+    boost::iostreams::back_insert_device<std::string> cmp(compressed);
+    boost::iostreams::stream<boost::iostreams::back_insert_device<std::string>> s_cmp(cmp);
+
+    boost::iostreams::filtering_ostreambuf out;
+    out.push(boost::iostreams::zlib_compressor());
+    out.push(boost::iostreams::base64_encoder());
+    out.push(s_cmp);
+    boost::iostreams::copy(s_src, out);
+    s_cmp.flush();
+
+    return compressed;
+}
+
+void Message::Decompress() noexcept {
+    if (m_is_compressed) {
+        std::string_view::size_type temp = m_message_size;
+        m_message_text = DecompressImpl(m_message_text);
+        m_message_size = m_message_text.size();
+        m_is_compressed = false;
+    }
+}
+
+std::string Message::DecompressImpl(const std::string& data) {
+    boost::iostreams::basic_array_source<char> src(data.c_str(), data.size());
+    boost::iostreams::stream<boost::iostreams::basic_array_source<char>> s_src(src);
+
+    std::string decompressed;
+    decompressed.reserve(std::pow(2.0, 26.0));
```

Comment:
would be better to send the message size and allocate appropriate space. maybe with a safety check / max message size to avoid trying to allocate too much.

Reply:
We have to send compressed size as well so network part will know where message ends.

Reply:
Changed reserve size to uncompressed size - as found in the received header - in https://github.com/freeorion/freeorion/pull/3854/commits/4ead6795ce2b51bce839955a679750840ada4831. If the reserve size is too large, an error is logged and an exception is thrown; I don't think there's much that can be done in this case.